File: 04-producing-consuming-kafka-with-java/src/java-kafka/src/main/java/com/trivadis/kafkaws/consumer/KafkaConsumerAuto.java
package com.trivadis.kafkaws.consumer;

import java.time.Duration;
import java.util.Collections;
import java.util.Properties;

import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.ConsumerRecords;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.common.serialization.LongDeserializer;
import org.apache.kafka.common.serialization.StringDeserializer;

public class KafkaConsumerAuto {

    private final static String TOPIC = "test-java-topic";
    private final static String BOOTSTRAP_SERVERS
            = "localhost:9092,localhost:9093,localhost:9094";
    private final static Duration CONSUMER_TIMEOUT = Duration.ofSeconds(1);

    private static Consumer<Long, String> createConsumer() {
        Properties props = new Properties();
        props.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, BOOTSTRAP_SERVERS);
        props.put(ConsumerConfig.GROUP_ID_CONFIG, "KakfaConsumerAuto");
        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, true);
        props.put(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG, 10000);
        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, LongDeserializer.class.getName());
        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());

        // Create the consumer using props.
        Consumer<Long, String> consumer = new KafkaConsumer<>(props);

        // Subscribe to the topic.
        consumer.subscribe(Collections.singletonList(TOPIC));
        return consumer;
    }

    private static void runConsumer(int waitMsInBetween) throws InterruptedException {
        final int giveUp = 100;

        try (Consumer<Long, String> consumer = createConsumer()) {
            int noRecordsCount = 0;

            while (true) {
                ConsumerRecords<Long, String> consumerRecords = consumer.poll(CONSUMER_TIMEOUT);

                if (consumerRecords.isEmpty()) {
                    noRecordsCount++;
                    if (noRecordsCount > giveUp) {
                        break;
                    } else {
                        continue;
                    }
                }

                consumerRecords.forEach(record -> {
                    System.out.printf("%d - Consumer Record:(Key: %d, Value: %s, Partition: %d, Offset: %d)\n",
                            consumerRecords.count(), record.key(), record.value(),
                            record.partition(), record.offset());
                    try {
                        // Simulate slow processing
                        Thread.sleep(waitMsInBetween);
                    } catch (InterruptedException e) {
                    }
                });
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        System.out.println("DONE");
    }

    public static void main(String... args) throws Exception {
        if (args.length == 0) {
            runConsumer(10);
        } else {
            runConsumer(Integer.parseInt(args[0]));
        }
    }

}

File: 06-producing-consuming-kafka-with-kotlin/src/kotlin-simple-kafka/src/main/kotlin/com/trivadis/kafkaws/consumer/KafkaConsumerAuto.kt
@file:JvmName("KafkaConsumerAuto")
package com.trivadis.kafkaws.consumer

import org.apache.kafka.clients.consumer.ConsumerConfig.*
import org.apache.kafka.clients.consumer.KafkaConsumer
import org.apache.kafka.common.serialization.LongDeserializer
import org.apache.kafka.common.serialization.StringDeserializer
import java.time.Duration
import java.util.*


private val TOPIC = "test-kotlin-topic"
private val BOOTSTRAP_SERVERS = "dataplatform:9092,dataplatform:9093"

fun runConsumer(waitMsInBetween: Long) {
    // Define properties.
    val props = Properties()
    props[BOOTSTRAP_SERVERS_CONFIG] = BOOTSTRAP_SERVERS
    props[GROUP_ID_CONFIG] = "kotlin-simple-consumer"
    //props[ENABLE_AUTO_COMMIT_CONFIG] = true
    props[AUTO_COMMIT_INTERVAL_MS_CONFIG] = 10000
    props[KEY_DESERIALIZER_CLASS_CONFIG] = LongDeserializer::class.qualifiedName
    props[VALUE_DESERIALIZER_CLASS_CONFIG] = StringDeserializer::class.qualifiedName

    val consumer = KafkaConsumer<Long, String>(props).apply {
        subscribe(listOf(TOPIC))
    }

    consumer.use {
        while(true) {
            val messages = consumer.poll(Duration.ofMillis(100))

            messages.forEach {
                println("Consumed record [Key: ${it.key()}, Value: ${it.value()}] @ Partition: ${it.partition()}, Offset:  ${it.offset()}")
            }

            // Simulate slow processing
            Thread.sleep(waitMsInBetween);
        }
    }
}

fun main(args: Array<String>) {
    if (args.size == 0) {
        runConsumer(10)
    } else {
        runConsumer(args[0].toLong())
    }
}
